### PR TITLE
Mod Databank entries.

### LINKED ISSUE #326: Mod Databank Entries system
**Describe the feature**
A databank entry for every mod that has the name, version, author and an optional description.

### PR DESCRIPTION
**Pull request by jonahnm**
### Changes made in this pull request

  - Handler to add databank entries for mods.
  
May need further testing.

### CODE REVIEW

**Comment by jonahnm** [comment]
Closes #326

**Comment by 32Kallies** [comment]
> Allows #326 to be closed.

Keep in mind that you can do this too: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    }
+
+    private static void CompleteRegister(PDAEncyclopedia.EntryData data)
+    {
+        if (isinit)
+        {
+            InternalLogger.Info($"{data.key} entry added.");
+            PDAHandler.AddEncyclopediaEntry(data);
+            PDAEncyclopedia.Add(data.key, true);
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Does not compile in Below Zero

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+namespace Nautilus.Handlers;
+public class ModDataBankHandler
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Should probably be Databank rather than DataBank

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    }
+    /// <summary>
+    /// Register mod with default values. 
+    /// </summary>
+    /// <param name="assembly">Assembly for your mod.</param>
+    public static void RegisterMod(Assembly assembly)
+    {
+        var bepinplugin = assembly.GetTypes().First(type => type.GetCustomAttributes(false).OfType<BepInPlugin>().Any());
+        var bepinplugindata = bepinplugin.GetCustomAttributes(false).OfType<BepInPlugin>().FirstOrDefault();
```
Comment (JKohlman) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Instead of using reflection off the assembly it might make more sense to just use the chainloader and have mods pass their GUID?

https://docs.bepinex.dev/v5.4.21/api/BepInEx.Bootstrap.Chainloader.html#BepInEx_Bootstrap_Chainloader_PluginInfos

File: Nautilus/Patchers/PDAEncyclopediaPatcher.cs [hunk 4]
```diff
@@ -1,5 +1,6 @@
-﻿using System.Collections.Generic;
+using System.Collections.Generic;
 using HarmonyLib;
+using Nautilus.Handlers;
```
Comment (JKohlman) [inline Nautilus/Patchers/PDAEncyclopediaPatcher.cs]: does this import need to be added?

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+using UnityEngine.SceneManagement;
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: That is a lot of using statements, are all of them used?

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 2 above)
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Missing class documentation, look at other handler classes for inspiration.

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            string version = "";
+            if(!data.version.IsNullOrWhiteSpace())
+            {
+                version = " " + data.version;
+            }
+            LanguageHandler.SetLanguageLine($"Ency_{data.guid}", data.name + version);
+        }
+        CompleteRegister(entrydata);
+    }
+    public record struct ModData
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Also missing documentation

File: Nautilus/Patchers/PDAEncyclopediaPatcher.cs (hunk 4 above)
Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaPatcher.cs]: No, this was leftover from previous modifications to that file that I removed, seems I forgot to remove the import.

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            PDAHandler.AddEncyclopediaEntry(data);
+            PDAEncyclopedia.Add(data.key, true);
+        }else
+        {
+            waitlist.Add(data);
+        }
+
+    }
+    /// <summary>
+    /// Register mod with default values.
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Could maybe be more descriptive... what does "register mod" mean?

File: Nautilus/Patchers/PDAEncyclopediaPatcher.cs (hunk 4 above)
Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaPatcher.cs]: Fixed.

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Nautilus.Patchers;
+internal class PDAEncyclopediaTabPatcher
+{
+    internal static void Patch(Harmony harmony)
+    {
+        harmony.PatchAll(typeof(PDAEncyclopediaTabPatcher));
+    }
+    [HarmonyPatch(typeof(uGUI_EncyclopediaTab))]
+    [HarmonyPatch(nameof(uGUI_EncyclopediaTab.Awake))]
```
Comment (EldritchCarMaker) [inline Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs:13]: This only works once, so mods can’t arbitrarily add/remove their entries at runtime. Try to change this if you can, ideally everything in nautilus can be used at any arbitrary point in runtime

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        var bepinplugin = assembly.GetTypes().First(type => type.GetCustomAttributes(false).OfType<BepInPlugin>().Any());
+        var bepinplugindata = bepinplugin.GetCustomAttributes(false).OfType<BepInPlugin>().FirstOrDefault();
+        var entrydata = new PDAEncyclopedia.EntryData()
+        {
+            image = null,
+            key = bepinplugindata.GUID,
+            unlocked = true,
+            path = "Mods",
+            nodes = PDAEncyclopedia.ParsePath("Mods")
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: What is this the call to ParsePath for? And also for all places related to registering, can't we use Nautilus's PDAHandler instead?

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 3 above)
Comment (EldritchCarMaker) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Yea, that’s probably better

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    private static void CompleteRegister(PDAEncyclopedia.EntryData data)
+    {
+        if (isinit)
+        {
+            InternalLogger.Info($"{data.key} entry added.");
+            PDAHandler.AddEncyclopediaEntry(data);
+            PDAEncyclopedia.Add(data.key, true);
+        }else
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: O_O

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 3 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: I feel that there wouldn't be much of a performance difference, but I'll do it anyways.

**Review by 32Kallies — CHANGES_REQUESTED** [review]

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 1 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: I don't own Below Zero, so I am not able to fix that.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 2 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Agreed.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 3 above)
Comment (JKohlman) [reply Nautilus/Handlers/ModDataBankHandler.cs]: In terms of performance, I imagine it would be a bit faster as it's a single dictionary lookup rather than an iteration of the types. But vastly more importantly, it's simpler IMO.

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        LanguageHandler.SetLanguageLine($"Ency_{bepinplugindata.GUID}", bepinplugindata.Name + " " + bepinplugindata.Version.ToString());
+        LanguageHandler.SetLanguageLine($"EncyDesc_{bepinplugindata.GUID}", "A BepInEx plugin using Nautilus.");
+        CompleteRegister(entrydata);
+    }
+    /// <summary>
+    /// Register mod with specified values.
+    /// </summary>
+    /// <param name="data"></param>
+    public static void RegisterMod(ModData data)
+    {
```
Comment (EldritchCarMaker) [inline Nautilus/Handlers/ModDataBankHandler.cs]: May be good to add a safety check here for mod guid/name. A simple string.IsNullOrEmpty should be good enough

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 3 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Should be fixed.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 1 above)
Comment (32Kallies) [reply Nautilus/Handlers/ModDataBankHandler.cs]: The NuGet packages provide all needed dependencies.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 2 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Resolved.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 5 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Fixed.

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 12]
```diff
@@ -43,10 +37,9 @@ private static void CompleteRegister(PDAEncyclopedia.EntryData data)
     /// Register mod with default values. 
     /// </summary>
     /// <param name="assembly">Assembly for your mod.</param>
-    public static void RegisterMod(Assembly assembly)
+    public static void RegisterMod(string GUID)
     {
-        var bepinplugin = assembly.GetTypes().First(type => type.GetCustomAttributes(false).OfType<BepInPlugin>().Any());
-        var bepinplugindata = bepinplugin.GetCustomAttributes(false).OfType<BepInPlugin>().FirstOrDefault();
+        var bepinplugindata = BepInEx.Bootstrap.Chainloader.PluginInfos.FirstOrDefault(info => info.Value.Metadata.GUID == GUID).Value.Metadata;
```
Comment (JKohlman) [inline Nautilus/Handlers/ModDataBankHandler.cs]: `BepInEx.Bootstrap.Chainloader.PluginInfos[GUID]`? Or `.TryGetValue` if you wanna be extra safe. No need to iterate a dictionary as it's meant to do lookups

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 7 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Made it a bit more descriptive.

**Review by JKohlman — CHANGES_REQUESTED** [review]

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 12 above)
Comment (JKohlman) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Probably should use the TryGetValue way tbh

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 10 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Fixed.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 12 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Ah alright.

Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Should be fixed.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 9 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: If I don't call it an error occurs.

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 13]
```diff
@@ -27,7 +27,7 @@ private static void CompleteRegister(PDAEncyclopedia.EntryData data)
             InternalLogger.Info($"{data.key} entry added.");
             PDAHandler.AddEncyclopediaEntry(data);
             PDAEncyclopedia.Add(data.key, true);
-        }else
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: image

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 9 above)
Comment (32Kallies) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Hmm, you can't just do `new string[] { "Mods" }`?

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 1 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Should be fixed.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 13 above)
Comment (32Kallies) [reply Nautilus/Handlers/ModDataBankHandler.cs]: I personally would start the else on a new line but IDK that's just me

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    [HarmonyPostfix]
+    internal static void EncyTabAwakePostfix(uGUI_EncyclopediaTab __instance)
+    {
+      ModDataBankHandler.Initialize(__instance);
```
Comment (EldritchCarMaker) [inline Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: This should also be a config option. Not everyone wants the mods to be listed in the encyclopedia, it should definitely be toggleable

**Review by EldritchCarMaker — CHANGES_REQUESTED** [review]

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs (hunk 8 above)
Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs:13]: They should be able to add at any point.

Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs:13]: The CompleteRegister function will work at any time in runtime.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 11 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: It isn't needed, the record struct has those two fields as required. (You can't create the struct without those set iirc)

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 15]
```diff
@@ -0,0 +1,111 @@
+using BepInEx;
+using Nautilus.Utility;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using UnityEngine;
+
+namespace Nautilus.Handlers;
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Class should be static

**Review by 32Kallies — CHANGES_REQUESTED** [review]

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 7 above)
Comment (32Kallies) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Maybe just me but I still don’t know what the method does when reading that.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 15 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: ah, you're right, was actually just thinking about that haha.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 7 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: I feel it's quite self-explanatory.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 2 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Added some docs.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 6 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Added docs.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 13 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: It is now in a new line.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 15 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: fixed.

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs (hunk 14 above)
Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: This isn't feasible as encyclopedia entries are stored in the save file.

File: Example mod/ConfigExamples.cs [hunk 16]
```diff
@@ -5,14 +5,14 @@
...
 using Nautilus.Utility;
-
 namespace Nautilus.Examples;
 
 using HarmonyLib;
 using Handlers;
 using UnityEngine;
 using BepInEx;
 using BepInEx.Logging;
+using System.Reflection;
```
Comment (JKohlman) [inline Example mod/ConfigExamples.cs]: Don't need anymore

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+            LanguageHandler.SetLanguageLine($"EncyDesc_{data.guid}", data.desc);
+        }
+        if(data.name != null)
+        {
+            string version = "";
+            if(!data.version.IsNullOrWhiteSpace())
+            {
+                version = " " + data.version;
+            }
+            LanguageHandler.SetLanguageLine($"Ency_{data.guid}", data.name + version);
```
Comment (JKohlman) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Already using string formatting so keep using it, something like

`$"Ency_{data.guid}", $"{data.name} {data.version ?? ""}"`

(also removes if statement)

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        var bepinplugindata = info.Metadata;
+        var entrydata = new PDAEncyclopedia.EntryData()
+        {
+            image = null,
+            key = bepinplugindata.GUID,
+            unlocked = true,
+            path = "Mods",
+            nodes = PDAEncyclopedia.ParsePath("Mods")
+        };
+        LanguageHandler.SetLanguageLine($"Ency_{bepinplugindata.GUID}", bepinplugindata.Name + " " + bepinplugindata.Version.ToString());
```
Comment (JKohlman) [inline Nautilus/Handlers/ModDataBankHandler.cs]: Use fully formatted strings instead of combining them with addition

`$"Ency_{bepinplugindata.GUID}",$"{bepinplugindata.Name} {bepinplugindata.Version}"`

File: Nautilus/Initializer.cs [hunk 19]
```diff
@@ -1,6 +1,7 @@
 using System;
 using BepInEx;
 using HarmonyLib;
+using Nautilus.Handlers;
```
Comment (JKohlman) [inline Nautilus/Initializer.cs]: Necessary?

File: Example mod/ConfigExamples.cs (hunk 16 above)
Comment (jonahnm) [reply Example mod/ConfigExamples.cs]: fixed.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 17 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: fixed.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 18 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: fixed.

File: Nautilus/Initializer.cs (hunk 19 above)
Comment (jonahnm) [reply Nautilus/Initializer.cs]: fixed.

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs (hunk 14 above)
Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: Implemented.

**Comment by jonahnm** [comment]
Review needed.

**Comment by 32Kallies** [comment]
After loading into the game, I do not see any mod databank entries or any tab for that.

**Comment by 32Kallies** [comment]
This seems to work as intended with mods that explicitly call the `RegisterMod` method, however I imagined it would automatically add mod databank entries for mods that don't. I'm fine either way, just something that came to mind.

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Nautilus.Handlers;
+/// <summary>
+/// A handler class for adding databank entries for mods.
+/// </summary>
+public static class ModDatabankHandler
+{
+    private static bool isinit;
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: naming conventions

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 21]
```diff
@@ -0,0 +1,111 @@
+using BepInEx;
+using Nautilus.Utility;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Nautilus.Handlers;
+/// <summary>
+/// A handler class for adding databank entries for mods.
+/// </summary>
+public static class ModDatabankHandler
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: file name does not match class name

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+{
+    private static bool isinit;
+    private static List<PDAEncyclopedia.EntryData> waitlist = new();
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: again, naming

**Comment by jonahnm** [comment]
ECM forced me to make it that way.

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs [hunk 23]
```diff
@@ -0,0 +1,24 @@
+using HarmonyLib;
+using Nautilus.Handlers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using UWE;
```
Comment (32Kallies) [inline Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: Unnecessary using statements

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Nautilus.Patchers;
```
Comment (32Kallies) [inline Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs:4]: there should be a newline below the namespace for consistency

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+            path = "Mods",
+            nodes = PDAEncyclopedia.ParsePath("Mods")
+        };
+        if(data.desc != null)
+        {
+            LanguageHandler.SetLanguageLine($"EncyDesc_{data.guid}", data.desc);
+        }
+        if(data.name != null)
+        {
+            string version = "";
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: what's this for?

File: Nautilus/Handlers/ModDataBankHandler.cs [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+            PDAEncyclopedia.Add(data.key, true);
+#endif
+        }
+        else
+        {
+            waitlist.Add(data);
+        }
+    }
+    /// <summary>
+    /// Register mod with database using default values.
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDataBankHandler.cs]: I still could not tell you what this method does based on the "Register mod with database" documentation, without looking into the source code. The same goes with the other method.

File: Nautilus/Handlers/PDAHandler.cs [hunk 27]
```diff
@@ -220,8 +220,7 @@ public static void AddEncyclopediaEntry(PDAEncyclopedia.EntryData entry)
         }
 
         PDAEncyclopediaPatcher.CustomEntryData[entry.key] = entry;
-        
-        if (uGUI.isMainLevel)
+        if(PDAEncyclopedia.initialized)
```
Comment (32Kallies) [inline Nautilus/Handlers/PDAHandler.cs:224]: Alright, this should really be in a different PR but looks like it should have been done a long time ago.

File: Nautilus/Patchers/OptionsPanelPatcher.cs [hunk 28]
```diff
@@ -91,6 +92,7 @@ internal static void AddTabs_Postfix(uGUI_OptionsPanel __instance)
         // Maybe this could be split into its own file to handle nautilus options, or maybe it could be removed alltogether
         optionsPanel.AddHeading(modsTab, "Nautilus");
         optionsPanel.AddToggleOption(modsTab, "Enable debug logs", Utility.InternalLogger.EnableDebugging, Utility.InternalLogger.SetDebugging);
+        optionsPanel.AddToggleOption(modsTab, "Enable Mod Databank Entries", ModDatabankHandler.isEnabled);
```
Comment (32Kallies) [inline Nautilus/Patchers/OptionsPanelPatcher.cs]: I noticed that this is the only Nautilus option that capitalizes every word as opposed to just the first (`Enable Mod Databank Entries` vs `Enable mod databank entries`)

**Review by 32Kallies — CHANGES_REQUESTED** [review]

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 26 above)
Comment (32Kallies) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Pretend the modder knows nothing about this method and write your documentation like that.

Comment (32Kallies) [reply Nautilus/Handlers/ModDataBankHandler.cs]: It's not like you need to list everything it does internally, but the user should know all of its effects and its purpose.

File: Nautilus/Handlers/PDAHandler.cs (hunk 27 above)
Comment (32Kallies) [reply Nautilus/Handlers/PDAHandler.cs:224]: But if you want it probably should be moved into a different PR so we can track such a major change more easily.

**Comment by 32Kallies** [comment]
> ECM forced me to make it that way.

I doubt that happened but it's probably fine without automatic patching since that would clutter the tab with empty mod entries.

File: Nautilus/Handlers/PDAHandler.cs (hunk 27 above)
Comment (32Kallies) [reply Nautilus/Handlers/PDAHandler.cs:224]: And honestly there's no "if you want" to it, really should be organized into a separate PR.

Comment (jonahnm) [reply Nautilus/Handlers/PDAHandler.cs:224]: Hmmm, this PR doesn't work without this change.

File: Nautilus/Patchers/OptionsPanelPatcher.cs (hunk 28 above)
Comment (jonahnm) [reply Nautilus/Patchers/OptionsPanelPatcher.cs]: I'm not very good at naming conventions haha

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 25 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: good question.

File: Nautilus/Handlers/ModDataBankHandler.cs (hunk 26 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDataBankHandler.cs]: Is it better now?

**Comment by jonahnm** [comment]
> > ECM forced me to make it that way.
> 
> I doubt that happened but it's probably fine without automatic patching since that would clutter the tab with empty mod entries.

Well, it's more they wanted me to.

File: Nautilus/Handlers/PDAHandler.cs (hunk 27 above)
Comment (jonahnm) [reply Nautilus/Handlers/PDAHandler.cs:224]: So they'd have to be merged at the same time.

Comment (jonahnm) [reply Nautilus/Handlers/PDAHandler.cs:224]: Done.

**Comment by jonahnm** [comment]
Depends on #400.

File: Example mod/ConfigExamples.cs [hunk 29]
```diff
@@ -5,7 +5,6 @@
 using Nautilus.Options;
 using Nautilus.Options.Attributes;
 using Nautilus.Utility;
-
```
Comment (32Kallies) [inline Example mod/ConfigExamples.cs]: Why would line spacing this be removed? More consistent to keep it.

File: Nautilus/Handlers/ModDatabankHandler.cs [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+/// </summary>
+public static class ModDatabankHandler
+{
+    private static bool isInit;
+    private static List<PDAEncyclopedia.EntryData> waitList = new();
+    internal static bool isEnabled = true;
+    internal static void Initialize(uGUI_EncyclopediaTab tab)
+    {
+        LanguageHandler.SetLanguageLine("EncyPath_Mods", "Mods");
+        isinit = true;
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDatabankHandler.cs]: Compiler error because the field was renamed.

File: Nautilus/Handlers/ModDatabankHandler.cs [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Nautilus.Handlers;
+/// <summary>
+/// A handler class for adding databank entries for mods.
+/// </summary>
+public static class ModDatabankHandler
+{
+    private static bool isInit;
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDatabankHandler.cs]: Does your IDE not warn you when it isn't named like `_isInit` or `_waitList`? It still isn't following the repository's enforced standards.

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs [hunk 32]
```diff
@@ -0,0 +1,20 @@
+using HarmonyLib;
+using Nautilus.Handlers;
+using System;
```
Comment (32Kallies) [inline Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: Unnecessary using statement.

File: Nautilus/Handlers/ModDatabankHandler.cs [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+    private static List<PDAEncyclopedia.EntryData> waitList = new();
+    internal static bool isEnabled = true;
+    internal static void Initialize(uGUI_EncyclopediaTab tab)
```
Comment (32Kallies) [inline Nautilus/Handlers/ModDatabankHandler.cs]: Does the `tab` parameter serve any purpose?

**Review by 32Kallies — CHANGES_REQUESTED** [review]
Almost there.

File: Nautilus/Handlers/ModDatabankHandler.cs (hunk 31 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDatabankHandler.cs]: oh, no it doesn't.

File: Nautilus/Handlers/ModDatabankHandler.cs (hunk 30 above)
Comment (jonahnm) [reply Nautilus/Handlers/ModDatabankHandler.cs]: ah.

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs [hunk 34]
```diff
@@ -0,0 +1,19 @@
...
+    internal static void Patch(Harmony harmony)
+    {
+        harmony.PatchAll(typeof(PDAEncyclopediaTabPatcher));
+    }
+    [HarmonyPatch(typeof(uGUI_EncyclopediaTab))]
+    [HarmonyPatch(nameof(uGUI_EncyclopediaTab.Awake))]
+    [HarmonyPostfix]
+    internal static void EncyTabAwakePostfix(uGUI_EncyclopediaTab __instance)
+    {
+      ModDatabankHandler.Initialize(__instance);
```
Comment (32Kallies) [inline Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: Compiler error... you do have an IDE right?

**Review by 32Kallies — CHANGES_REQUESTED** [review]
So close.

File: Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs (hunk 34 above)
Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: yeah, I was just doing this from github file edits since I wasn't at home.

Comment (jonahnm) [reply Nautilus/Patchers/PDAEncyclopediaTabPatcher.cs]: fixed.

**Review by 32Kallies — APPROVED** [review]
Looks good to me.
